src/decode/decode_gps_l2c.c

```diff
@@ -0,0 +1,133 @@
+#include <libswiftnav/cnav_msg.h>
+#include <assert.h>
+
+#include "ephemeris.h"
+#include "track.h"
+#include "sbp.h"
+#include "sbp_utils.h"
+#include "signal.h"
+
+/** GPS L2C CNAV message length in symbols */
+#define GPS_CNAV_MSG_LENGTH         (300)
+
+/** GPS L2C symbol length [ms] */
+#define GPS_L2C_SYMBOL_LENGTH          20
+
+/** Maximum probability of symbol 0 as
+    expected by cnav_msg_decoder_add_symbol(). */
+#define MAX_PROBABILITY_OF_SYMBOL_0        0
+
+/** Maximum probability of symbol 1 as
+    expected by cnav_msg_decoder_add_symbol(). */
+#define MAX_PROBABILITY_OF_SYMBOL_1        0xFF
+
+typedef struct {
+  cnav_msg_t cnav_msg;
+  cnav_msg_decoder_t decoder;
+} gps_l2c_decoder_data_t;
+
+static decoder_t gps_l2c_decoders[NUM_GPS_L2CM_DECODERS];
+static gps_l2c_decoder_data_t gps_l2c_decoder_data[NUM_GPS_L2CM_DECODERS];
+
+static void decoder_gps_l2c_init(const decoder_channel_info_t *channel_info,
+                                 decoder_data_t *decoder_data);
+static void decoder_gps_l2c_disable(const decoder_channel_info_t *channel_info,
+                                    decoder_data_t *decoder_data);
+static void decoder_gps_l2c_process(const decoder_channel_info_t *channel_info,
+                                    decoder_data_t *decoder_data);
+
+static const decoder_interface_t decoder_interface_gps_l2c = {
+  .code =         CODE_GPS_L2CM,
+  .init =         decoder_gps_l2c_init,
+  .disable =      decoder_gps_l2c_disable,
+  .process =      decoder_gps_l2c_process,
+  .decoders =     gps_l2c_decoders,
+  .num_decoders = NUM_GPS_L2CM_DECODERS
+};
+
+static decoder_interface_list_element_t list_element_gps_l2c = {
+  .interface = &decoder_interface_gps_l2c,
+  .next = 0
+};
+
+void decode_gps_l2c_register(void)
+{
+    for (u32 i=0; i<NUM_GPS_L2CM_DECODERS; i++) {
+    gps_l2c_decoders[i].active = false;
+    gps_l2c_decoders[i].data = &gps_l2c_decoder_data[i];
+  }
+
+  decoder_interface_register(&list_element_gps_l2c);
+}
+
+static void decoder_gps_l2c_init(const decoder_channel_info_t *channel_info,
+                                 decoder_data_t *decoder_data)
+{
+  (void)channel_info;
+  gps_l2c_decoder_data_t *data = decoder_data;
+  cnav_msg_decoder_init(&data->decoder);
+}
+
+static void decoder_gps_l2c_disable(const decoder_channel_info_t *channel_info,
+                                    decoder_data_t *decoder_data)
+{
+  (void)channel_info;
+  (void)decoder_data;
+}
+
+static void decoder_gps_l2c_process(const decoder_channel_info_t *channel_info,
+                                    decoder_data_t *decoder_data)
+{
+  gps_l2c_decoder_data_t *data = decoder_data;
+
+  /* Process incoming nav bits */
+  s8 soft_bit;
+  while (tracking_channel_nav_bit_get(channel_info->tracking_channel,
+                                      &soft_bit)) {
+    /* Update TOW */
+    u8 symbol;
+    u32 delay;
+    s32 tow_ms;
+
+    symbol = (soft_bit >= 0) ? MAX_PROBABILITY_OF_SYMBOL_1 :
```

Comment:
You are right. My original approach was too harsh. I have changed the code. It should be fixed now.